packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,216 @@
+        .invoke({
+          issuer: id,
+          audience: id,
+          with: id.toDIDKey(),
+          nb: {
+            message: messageLink,
+          },
+          expiration: Infinity,
+        })
+        .delegate()
+      allocateUcanConcludefx.attach({
+        bytes,
+        cid: messageLink,
+      })
+
+      // Create result object
+      /** @type {API.OkBuilder<API.BlobAddSuccess, API.BlobAddFailure>} */
+      const result = Server.ok({
+        location: {
+          'ucan/await': ['.out.ok.claim', acceptfx.link()],
+        },
+      })
+
+      // In case blob allocate provided an address to write
+      // the blob is still not stored
+      if (blobAllocateOutAddress) {
+        const blobPut = HTTP.put.invoke({
+          issuer: putSubject,
+          audience: putSubject,
+          with: putSubject.toDIDKey(),
+          nb: {
+            blob,
+            address: blobAllocateOutAddress,
+          },
+          facts,
+          expiration: Infinity,
+        })
+
+        const putfx = await blobPut.delegate()
+
+        // store `http/put` invocation
+        // TODO: store implementation
+        // const archiveDelegationRes = await putfx.archive()
+        // if (archiveDelegationRes.error) {
+        //   return {
+        //     error: archiveDelegationRes.error
+        //   }
+        // }
+        const invocationPutRes = await tasksStorage.put(putfx)
+        if (invocationPutRes.error) {
+          return {
+            error: invocationPutRes.error,
+          }
+        }
+
+        return (
+          result
+            // 1. System attempts to allocate memory in user space for the blob.
+            .fork(allocatefx)
+            .fork(allocateUcanConcludefx)
+            // 2. System requests user agent (or anyone really) to upload the content
+            // corresponding to the blob
+            // via HTTP PUT to given location.
+            .fork(putfx)
+            // 3. System will attempt to accept uploaded content that matches blob
+            // multihash and size.
+            .join(acceptfx)
+        )
+      }
+
+      // Add allocate receipt if allocate was executed
+      if (allocateUcanConcludefx) {
+        return (
+          result
+            // 1. System attempts to allocate memory in user space for the blob.
+            .fork(allocatefx)
+            .fork(allocateUcanConcludefx)
+            // 3. System will attempt to accept uploaded content that matches blob
+            // multihash and size.
+            .join(acceptfx)
+        )
+      }
+
+      // Blob was already allocated and is already stored in the system
+      return (
+        result
+          // 1. System allocated memory in user space for the blob.
+          .fork(allocatefx)
+          .fork(allocateUcanConcludefx)
+          // 3. System will attempt to accept uploaded content that matches blob
+          // multihash and size.
+          .join(acceptfx)
+      )
+    },
+  })
```

Comment:
there is also other unknown, the case where other user already allocated and wrote, but you did not.

For me until now, I always expected the `receipt.out` to need to be the same, but thought the effects could vary (like in this case)

Reply:
also we cannot attach blocks if there is no link for it in the `nb's`. I think we are getting extremely complex with this flow now and I would much rather simplify instead of keeping more and more complexity